File: src/components/cytoscape/cytoscape.py
from typing import Any, Callable, Optional

from nicegui.element import Element
from nicegui import ui,app

import json
import logging

from models.graph import GraphService

class Cytoscape(Element, component='cytoscape.js'):
    """
    This component is based on https://js.cytoscape.org/
    """

    def __init__(
          self,
          title: str,
          width = None,
          height = None,
          graph = None
        ):
        super().__init__()

        # store graph id
        self.graph = graph

        self._props['title'] = title
        ui.add_head_html('<script src="https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.21/lodash.min.js"></script>')
        ui.add_head_html('<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>')
        ui.add_head_html('<script src="https://unpkg.com/cytoscape/dist/cytoscape.min.js"></script>')
        ui.add_head_html('<script src="/static/cytoscape-edgehandles.js"></script>')
        ui.add_head_html('<script src="/static/cytoscape-snap-to-grid.js"></script>')
        ui.add_head_html('<script src="/static/cytoscape-automove.js"></script>')
        ui.add_head_html(
            f'''
<style>
      .cy {{
        width: {width}px;
        height: {height}px;
        z-index: 999;
      }}
</style>
            ''')

        self.on('event', self.handle_event)
        self.on('edgehandle', self.handle_eventedgehandle)
        self.on('updateNodePosition', self.updateNodePosition)

        self.data_node = {
            "label": "",
            "reference": "",
            "alias": "",
            "group": "",
            "tag": ""
        }
        self.data_node_id = None

        self.data_edge = {
            "label": "",
            "reference": "",
        }
        self.data_edge_id = None

        # select dialog node
        self.dialog_node = ui.dialog()
        with self.dialog_node, ui.card() as card:
            card.style('width: 15%')
            card.style('max-width: 15%')
            card.style('height: 85%')
            card.style('max-height: 85%')

            ui.label('Node')

            for field in ['label','reference','alias','group','tag']:
              label = ui.input(label=field, placeholder='start typing value for {}'.format(field)).style('width: 100%')
              label.bind_value(self.data_node, target_name = field)
            
            ui.separator()

            ui.button('Alias', on_click=lambda: self.findAlias(self.data_node_id))
            ui.button('Drop', on_click=lambda: self.dropNode(self.data_node_id))
            ui.button('Clone', on_click=lambda: self.cloneNode(
              {
                "label": self.data_node["label"],
                "reference": self.data_node["reference"],
                "alias": self.data_node["alias"],
                "group": self.data_node["group"],
                "tag": self.data_node["tag"],
                "x": self.data_node["x"] + 50,
                "y": self.data_node["y"] + 50
              }, self.graph))
            ui.button('Draw', on_click=lambda: self.start())
            ui.button('Save', on_click=lambda: self.saveNode())

        # select dialog node
        self.dialog_edge = ui.dialog()
        with self.dialog_edge, ui.card():
            ui.label('Edge')

            for field in ['label','reference']:
              label = ui.input(label=field, placeholder='start typing value for {}'.format(field)).style('width: 100%')
              label.bind_value(self.data_edge, target_name = field)

            ui.button('Drop', on_click=lambda: self.dropEdge(self.data_edge_id))

    def handle_event(self, event):
      if event.args["type"] not in ['mouseout','mouseover']:
        logging.info(json.dumps(event.args))

      if event.args['type'] == 'click' and event.args['target']['type'] == 'node':
        for field in ['label','reference','alias','group','tag']:
          if field in event.args['target']['data']:
            self.data_node[field] = event.args['target']['data'][field]
        for field in ['x','y']:
          if field in event.args['target']['position']:
            self.data_node[field] = event.args['target']['position'][field]
        self.data_node_id = event.args['target']['id']
        self.dialog_node.open()

      if event.args['type'] == 'click' and event.args['target']['type'] == 'edge':
        for field in ['label','reference']:
          if field in event.args['target']['data']:
            self.data_edge[field] = event.args['target']['data'][field]

        self.data_edge_id = event.args['target']['id']
        self.dialog_edge.open()

    def handle_eventedgehandle(self, event):
      """
      Handle edgehandle events
      Cf. https://github.com/cytoscape/cytoscape.js-edgehandles
      """
      logging.info(json.dumps(event.args))

      if event.args['type'] == "ehcomplete":
        GraphService().addEdge(
            source = event.args['addedEdge']['source'],
            target = event.args['addedEdge']['target']
          )
        graph = GraphService().graph(id = self.graph)
        self.run_method("loadNodes", graph)

    def loadNodes(self, graph):
      """
      reload graph (nodes and edges)
      """
      self.run_method("loadNodes", graph)

    def loadStyle(self, graph):
      self.run_method("loadStyle", graph)

    def saveNode(self):
      """
      update node fields
      """
      GraphService().updateNode(id = self.data_node_id, data = {
              "label": self.data_node["label"],
              "reference": self.data_node["reference"],
              "alias": self.data_node["alias"],
              "group": self.data_node["group"],
              "tag": self.data_node["tag"]
            })
      graph = GraphService().graph(id = self.graph)
      self.run_method("loadNodes", graph)
      for key in self.data_node:
        self.data_node[key] = None
      self.dialog_node.close()

    def start(self):
      """
      start linking two nodes with edgehandle extention
      """
      self.run_method('start', self.data_node_id)
      self.dialog_node.close()

    def drawMode(self, value):
      """
      switch draw mode with edgehandle extention
      """
      self.run_method('drawMode', value)

    def groupMode(self, value):
      """
      activate/disable group mode
      Cf. https://github.com/cytoscape/cytoscape.js-automove/blob/master/cytoscape-automove.js
      """
      groups = GraphService().getGroups(self.graph)
      self.run_method('groupMode', value, groups)

    def select(self, data):
       """
       Select a single node onto current raph
       """
       self.run_method('select', data)

    def findAlias(self, id):
      selected = GraphService().getAlias(id)
      self.select(selected)

    def dropNode(self, id):
      GraphService().dropNode(id)
      self.run_method('dropNode', id)
      self.dialog_node.close()

    def dropEdge(self, id):
      GraphService().dropEdge(id)
      self.run_method('dropEdge', id)
      self.dialog_edge.close()

    def cloneNode(self, data, graph):
      cloned = GraphService().cloneNode(clone = data, id = graph)
      self.run_method('cloneNode', cloned)
      self.dialog_node.close()

    def getNodes(self):
      self.run_method('updateNodePosition')

    def fit(self):
      self.run_method('fit')

    async def png(self):
      png = await self.run_method('png')
      GraphService().updatePngToGraphById(self.graph, png)

    def updateNodePosition(self, event):
       for node in event.args:
          GraphService().updateNodePosition(node['data']['id'], node['position']['x'], node['position']['y'])

File: src/components/cytoscape/cytoscape.js
export default {
  mounted: function () {
    if (!this.container) {
      // builld container
      this.container = window.container = cytoscape({
        container: document.getElementById("container"),
        layout: { name: 'preset' },
        motionBlur: true,
        selectionType: 'single',
        boxSelectionEnabled: false
      });

      // Add cytoscape edgehandles
      this.edgehandles = this.container.edgehandles({
        canConnect: function (sourceNode, targetNode) {
          // whether an edge can be created between source and target
          return !sourceNode.same(targetNode); // e.g. disallow loops
        },
        edgeParams: function (sourceNode, targetNode) {
          // for edges between the specified source and target
          // return element object to be passed to cy.add() for edge
          return {};
        },
        hoverDelay: 150, // time spent hovering over a target node before it is considered selected
        snap: true, // when enabled, the edge can be drawn by just moving close to a target node (can be confusing on compound graphs)
        snapThreshold: 50, // the target node must be less than or equal to this many pixels away from the cursor/finger
        snapFrequency: 15, // the number of times per second (Hz) that snap checks done (lower is less expensive)
        noEdgeEventsInDraw: true, // set events:no to edges during draws, prevents mouseouts on compounds
        disableBrowserGestures: true // during an edge drawing gesture, disable browser gestures such as two-finger trackpad swipe and pinch-to-zoom      
      })
      // Add cytoscape snapToGrid
      this.snapToGrid = this.container.snapToGrid({
        gridSpacing: 100
      })
      // Add cytoscape automove
      this.automove = this.container.automove({
      })

      const emitNode = (evt) => {
        this.$emit("event", {
          type: evt.type,
          target: {
            id: evt.target.id(),
            data: evt.target.data(),
            position: {
              x: evt.target.position().x,
              y: evt.target.position().y
            },
            type: 'node'
          }
        });
      }

      const emitEdge = (evt) => {
        this.$emit("event", {
          type: evt.type,
          target: {
            id: evt.target.id(),
            data: evt.target.data(),
            type: 'edge'
          }
        });
      }

      const emitEdgeHandle = (evt, sourceNode, targetNode, addedEdge) => {
        this.$emit("edgehandle", {
          type: evt.type,
          sourceNode: sourceNode.data(),
          targetNode: targetNode.data(),
          addedEdge: addedEdge.data()
        });
      }

      this.container.on('zoom', (event) => {
      });

      this.container.on('dblclick', 'node', (event) => {
        emitNode(event)
      });

      this.container.on('dblclick', 'edge', (event) => {
        emitEdge(event)
      });

      this.container.on('cxttap', 'node', (event) => {
        emitNode(event)
      });

      this.container.on('cxttap', 'edge', (event) => {
        emitEdge(event)
      });

      this.container.on('click', (event) => {
      });

      this.container.on('click', 'node', (event) => {
        emitNode(event)
      });

      this.container.on('click', 'edge', (event) => {
        emitEdge(event)
      });

      this.container.on('mouseover', 'node', (event) => {
        emitNode(event)
      });

      this.container.on('mouseover', 'edge', (event) => {
        emitEdge(event)
      });

      this.container.on('mouseout', 'node', (event) => {
        emitNode(event)
      });

      this.container.on('mouseout', 'edge', (event) => {
        emitEdge(event)
      });

      this.container.on('ehcomplete', (event, sourceNode, targetNode, addedEdge) => {
        emitEdgeHandle(event, sourceNode, targetNode, addedEdge)
      });

      // Rules
      this.rules = []
    }

  },
  methods: {
    /**
     * load nodes
     * @param {*} cy 
     * @param {*} mygraph 
     */
    loadNodes(mygraph) {
      let cy = this.container

      // Disable existing rules
      _.each(this.rules, (rule) => {
        rule.destroy()
      })

      // refresh render
      cy.startBatch()
      cy.nodes().remove()
      cy.edges().remove()

      if (mygraph.nodes) {
        const mynodes = _.map(mygraph.nodes, (node) => {
          return {
            data: {
              id: node.id,
              label: node.label,
              reference: node.reference,
              cdata: node.cdata,
              alias: node.alias,
              group: node.group,
              tag: node.tag
            },
            position: {
              x: node.x,
              y: node.y
            }
          }
        })
        cy.add(mynodes)
      }

      if (mygraph.edges) {
        const myedges = _.map(mygraph.edges, (edge) => {
          return {
            data: {
              id: edge.id,
              label: edge.label,
              reference: edge.reference,
              source: edge.source,
              target: edge.target,
              _source: edge._source,
              _target: edge._target,
              tag: edge.tag
            }
          }
        })
        cy.add(myedges)
      }

      cy.endBatch()
      cy.fit()
    },
    /**
     * load style
     * @param {*} cy 
     * @param {*} mygraph 
     */
    loadStyle(mygraph) {
      let cy = this.container

      // refresh render
      cy.startBatch()

      const content = (element) => {
        let cdata = element.data().cdata ? '*' : ''
        let alias = element.data().alias ? '@' : ''
        if (element.data().label && element.data().group) {
          return `${element.data().label} (${element.data().group}) ${cdata} ${alias}`
        }
        if (element.data().label) {
          return `${element.data().label} ${cdata} ${alias}`
        }
        return ""
      }

      const mystyle = _.map(mygraph.styles, (tag) => {
        if (!tag.label) {
          let result = {
            selector: `${tag.selector}`,
            style: tag.style,
          }
          result.style.content = content
          return result
        } else {
          let result = {
            label: tag.label,
            selector: `${tag.selector}[tag = '${tag.label}']`,
            style: tag.style
          }
          result.style.content = content
          return result
        }
      })

      // Build style
      cy.style(mystyle)

      cy.endBatch()
      cy.fit()
    },
    groupMode(enable, groups) {
      if (enable) {
        // Enable all group
        _.each(this.rules, (rule) => {
          rule.destroy()
        })

        this.rules = []

        _.each(groups, (group) => {
          let selectedGroup = this.container.$(`[group = "${group}"]`)
          this.rules.push(container.automove({
            nodesMatching: selectedGroup,
            reposition: 'drag',
            dragWith: selectedGroup,
          }))
        })
      } else {
        // Disable group
        _.each(this.rules, (rule) => {
          rule.destroy()
        })
        this.rules = []
      }
    },
    select(data) {
      this.container.center(this.container.$(`#${data}`))
    },
    /**
   * drop node
   * @param {*} data 
   */
    dropNode(id) {
      this.container.$(`#${id}`).remove()
    },
    dropEdge(id) {
      this.container.$(`#${id}`).remove()
    },
    cloneNode(data) {
      let cloned = {
        data: {
          id: data.id,
          label: data.data?.label,
          reference: data.data?.reference,
          alias: data.data?.alias,
          cdata: data.data?.cdata,
          group: data.data?.group,
          tag: data.data?.tag
        },
        position: {
          x: data.position?.x,
          y: data.position?.y
        }
      }
      this.container.add(cloned)
    },
    updateNodePosition() {
      this.$emit("updateNodePosition", _.map(this.container.nodes(), (node) => {
        return {
          data: node.data(),
          position: node.position()
        }
      }));
    },
    /**
     * start draw mode
     * @param {*} data 
     */
    start(data) {
      let selected = this.container.$(`#${data.id}`)
      this.container.center(selected)
      this.edgehandles.start(selected)
    },
    /**
     * fi draw mode behaviour
     * @param {*} enable 
     */
    drawMode(enable) {
      if (enable) {
        this.edgehandles.enableDrawMode()
      } else {
        this.edgehandles.disableDrawMode()
      }
    },
    fit() {
      let cy = this.container
      cy.fit()
    },
    png() {
      let cy = this.container
      return cy.png({
        bg: 'white',
        full: this.preferences.full,
        maxWidth: this.preferences.maxWidth ? Number(this.preferences.maxWidth) : 1024,
        maxHeight: this.preferences.maxHeight ? Number(this.preferences.maxHeight) : 768
      }) || ""
    }
  },
  template: `
  <div class="cy">
    <div id="container" class="cy">
    </div>
  </div>
  `,
  data() {
    return {
      preferences: {
        full: false,
        maxWidth: 1024,
        maxHeight: 768
      }
    }
  },
  props: {
    title: String,
  },
};